File: app/src/main/java/com/liadpaz/amp/viewmodels/Playlist.java
package com.liadpaz.amp.viewmodels;

import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.Arrays;

public class Playlist implements Parcelable {
    public static final Creator<Playlist> CREATOR = new Creator<Playlist>() {
        @NonNull
        @Override
        public Playlist createFromParcel(@NonNull Parcel in) {
            return new Playlist(in);
        }

        @NonNull
        @Override
        public Playlist[] newArray(int size) {
            return new Playlist[size];
        }
    };

    public final String name;
    public final ArrayList<Song> songs;

    public Playlist(@NonNull String name, @NonNull ArrayList<Song> songs) {
        this.name = name;
        this.songs = songs;
    }

    private Playlist(@NonNull Parcel in) {
        name = in.readString();
        songs = in.createTypedArrayList(Song.CREATOR);
    }

    @Override
    public boolean equals(@Nullable Object obj) {
        if (obj instanceof Playlist) {
            Playlist other = (Playlist)obj;
            return name.equals(other.name) && songs.equals(other.songs);
        }
        return false;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(@NonNull Parcel dest, int flags) {
        dest.writeString(name);
        dest.writeTypedList(songs);
    }

    @NonNull
    @Override
    public String toString() {
        return "name: " + name + "\nsongs: " + Arrays.deepToString(songs.toArray());
    }
}

File: app/src/main/java/com/liadpaz/amp/viewmodels/Playlist.kt
package com.liadpaz.amp.viewmodels

import android.os.Parcel
import android.os.Parcelable
import java.util.*

data class Playlist(val name: String, val songs: ArrayList<Song>) : Parcelable {

    private constructor(parcel: Parcel) : this(parcel.readString()!!, parcel.createTypedArrayList(Song.CREATOR)!!)

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(name)
        dest.writeTypedList(songs)
    }

    override fun toString(): String {
        return """
            name: $name
            songs: ${songs.toTypedArray().contentDeepToString()}
            """.trimIndent()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (this.javaClass != other?.javaClass) return false
        other as Playlist
        return other.name == name && other.songs == songs
    }

    override fun hashCode(): Int {
        var result = name.hashCode()
        result = 31 * result + songs.hashCode()
        return result
    }

    companion object {
        val CREATOR: Parcelable.Creator<Playlist> = object : Parcelable.Creator<Playlist> {
            override fun createFromParcel(`in`: Parcel): Playlist {
                return Playlist(`in`)
            }

            override fun newArray(size: Int): Array<Playlist?> {
                return arrayOfNulls(size)
            }
        }
    }
}